background images based on embedded images and data-url #1175

speckyspooky opened the pull request:
**1. Enhancement of background image handling** 

The change includes the enhancemet that BIRT can handle background images based on the 3 source types
- URL
- Data-URL
- Report embedded images

The new option is given for:
- MasterPage
- Grid
- Table
- Row
- Cell
- TextElements
- all elements wich use backgroundImage

The main changes are given on the "BackgroundImageInfo.java" and "BackgroundImageType.java".
Important is that I enhanced the constructors to but the necessary information to the background-image-object
to have all information to create the background image.

In addition to it I make the property "backgroundImageType" accessable which was original design
but not full implemented to use it at the background image side.

In addition to it I enhanced the designer to display the images.
On emitter side I enhanced the HTML-emitter and the PDF-emitter to show the images.

Test cases of the change - done with the all-in-one-launch:
main test 1: *MasterPages* with all 3 types of background image source
main test 2: *grid with grid-row-cell* background images with all 3 types of background image source
main test 3: different *text elements* with all 3 types of background image source
main test 4: "data table with JavaScript image setting" with all 3 types of background image source

Attached you will find my test reports: [images.zip](https://github.com/eclipse/birt/files/10429573/images.zip)


 **2. Fixing of warning / java documentation**
Many of these classes had lot of warnings based on missing java-doc e.g. the constant-classes
and also the interface-classes. I added the according comments.
I solved also a small part of the warnings based on constant-usage and the render-option-hints on the HTML-emitter.

wimjongman commented on UI/org.eclipse.birt.report.designer.core/src/org/eclipse/birt/report/designer/util/ImageManager.java:
```diff
@@ -50,9 +51,11 @@ public class ImageManager {
 	private static final ImageManager instance = new ImageManager();
 
-	private List invalidUrlList = new ArrayList();
+	private List<String> invalidUrlList = new ArrayList<String>();
```
new ArrayList<>();

wimjongman reviewed (commented):
I started the review but then I found that there are a lot of unrelated changes in this PR.

Please split up the PR in two.

1 PR where you add new code for the embedded image only.
2 PR where you do the code refactoring and adding of the javadoc.

speckyspooky commented:
Could you provide me the committer-link please?
(Yes, I signed only the ECA.)

Could you provide my some hints on it how can I split the PR?
I could only split based on class level because I added on each class which I changed the missing comments too e.g. CSSConstantValue.

Yes I know and see that I have done many changes but the only why would be to clean up the whole PR (= close it).
Remove all changes incl. comments (full rollback).
And afterwards 2 single PRs with the changes of background images and another PR with the comments.
It would be a many stuff to do it so.

wimjongman commented on model/org.eclipse.birt.report.model/src/org/eclipse/birt/report/model/simpleapi/Style.java:
```diff
@@ -97,6 +108,40 @@ public void setBackgroundImage(String value) throws SemanticException {
+		return style.getBackgroundImageType();
+	}
+
+	/*
```
remove non-javadoc

wimjongman reviewed (changes requested):
Hi @speckyspooky. I have reviewed the code on sanity and it all looks good to me. I have a couple of small changes. 

It is very hard to review such a change. So in the future, we should split functionality from source cleanup and make two separate PR's.

Thanks for the updated comments.

Please update the minor things and I will accept this patch.

speckyspooky commented:
@wimjongman I have reworked and added your requested changes.

And of course, the next PR will be splitted between changes and refactoring/comment changes
for a better review and easier understanding of all.

wimjongman commented on model/org.eclipse.birt.report.model/src/org/eclipse/birt/report/model/simpleapi/Style.java:
```diff
@@ -72,27 +70,21 @@ public void setBackgroundAttachment(String value) throws SemanticException {
 
-	/*
-	 * (non-Javadoc)
+	/**
```
actually I meant to remove the javadoc completely since it should be documented in the super

wimjongman reviewed (approved):
Looks good to me. We wait for the build.

wimjongman commented:
Thanks, @speckyspooky !!

speckyspooky commented:
@wimjongman Thanks for your verification. I tested the latest Snapshot with my test reports and they work all together without any issue.
